Feature reduced hierarchy transform update

rams3s opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fishingcactus/openfl/118)
<!-- Reviewable:end -->

Commit:
Get rid of renderScaleTransform

Commit:
Merge remote-tracking branch 'origin/develop' into feature-reduced_hierarchy_transform_update

Commit:
Fixed undefined graphics owner when using format.swf.lite.MovieClip.__createShape()

Commit:
Removed unused DisplayObject.__getTransformedRenderBounds()

Commit:
Use render transform base stack to avoid updating the whole hierarchy transforms

Commit:
Fixed 9 slice rendering for render transform stack

Commit:
Fixed masks not being rendered

NuclearCookie committed:
Merge remote-tracking branch 'origin/develop' into feature-reduced_hierarchy_transform_update

NuclearCookie commented on format/swf/lite/MovieClip.hx:
```diff
@@ -710,14 +710,13 @@ class MovieClip extends flash.display.MovieClip {
+	private function __update9SliceBitmap ():Void {
 
 		// :TODO: should be in a prerender phase
 		// :TODO: use dirty flag if need to update __9SliceBitmap
```
you adressed these todos

NuclearCookie commented on openfl/display/BitmapData.hx:
```diff
@@ -412,7 +422,7 @@ class BitmapData implements IBitmapDrawable {
 		if (colorTransform != null) {
 			
-			var copy = new BitmapData (Reflect.getProperty (source, "width"), Reflect.getProperty (source, "height"), true, 0);
+			var copy = new BitmapData (Reflect.getProperty (source, "__width"), Reflect.getProperty (source, "__height"), true, 0);
```
use Reflect.getField now instead of getProperty

rams3s replied:
true

NuclearCookie commented on openfl/display/BitmapData.hx:
```diff
@@ -496,7 +506,7 @@ class BitmapData implements IBitmapDrawable {
 		if (colorTransform != null) {
 			
-			var copy = new BitmapData (Reflect.getProperty (source, "width"), Reflect.getProperty (source, "height"), true, 0);
+			var copy = new BitmapData (Reflect.getProperty (source, "__width"), Reflect.getProperty (source, "__height"), true, 0);
```
same here

rams3s replied:
true again

NuclearCookie commented:
Also, When reusing a 9slice graphic, I am experiencing issues with transparency. ( state does not seem to be reset ).

NuclearCookie commented:
Blend modes also seem broken ( multiply )

NuclearCookie commented on format/swf/lite/MovieClip.hx:
```diff
@@ -710,14 +710,13 @@ class MovieClip extends flash.display.MovieClip {
 		// :TODO: should be in a prerender phase
 		// :TODO: use dirty flag if need to update __9SliceBitmap
 
 		if (__symbol != null && __symbol.scalingGridRect != null && __9SliceBitmap == null) {
```
This condition can be avoided if you look at the place this function is called

rams3s replied:
not exactly the same test

NuclearCookie commented on openfl/_internal/renderer/RenderSession.hx:
```diff
@@ -42,13 +43,57 @@ class RenderSession {
+		pushRenderTargetBaseTransform (@:privateAccess Matrix.__identity, null);
 	}
 	
+	public function pushRenderTargetBaseTransform (source:Dynamic, renderTargetBaseTransform:Matrix) {
```
Source could be IBitmapDrawable

rams3s replied:
done

NuclearCookie commented on format/swf/lite/MovieClip.hx:
```diff
@@ -736,16 +736,14 @@ class MovieClip extends flash.display.MovieClip {
+				renderTransform.translate(-bounds.x, -bounds.y);
+				@:privateAccess bitmap.__drawGL(renderSession, this, renderTransform);
+				Matrix.pool.put (renderTransform);
 
```
re-add the bounds to the pool

rams3s replied:
good catch ;)

NuclearCookie commented:
I am experiencing issues with cached as bitmap. If I use it, masks seem to be ignored. 
First thought:
When drawing the mask, we used to update all the children https://github.com/FishingCactus/openfl/pull/118/files#diff-2b58ec07f66353241b751100e39589a9L181
This will have to be added somewhere

Commit:
Fixed Rectangle not being released to pool in 9 slice bitmap + some reviews

NuclearCookie committed:
Merge remote-tracking branch 'origin/develop' into feature-reduced_hierarchy_transform_update

Commit:
Merge remote-tracking branch 'origin/feature/colortransform_pool' into feature-reduced_hierarchy_transform_update

Commit:
Fixed square alpha issue

Commit:
Added renderAlpha and renderColorTransform to avoid having to reupdate all sub hierarchy when caching as bitmap.

This also fixes square alpha being applied on cached display object. At the same time, blend mode and color tansform arguments to GLBitmap.drawBitmapDrawable() that were never used have been removed as they would require a hierarchy update.

NuclearCookie committed:
Update the mask's cached bitmap if the flag __updateCachedBitmap is true

NuclearCookie committed:
Use correct bounds size to create cached bitmap

NuclearCookie committed:
reuse the existing cached bitmap instead of leaking it

NuclearCookie committed:
better mask transforms. almost correct

NuclearCookie committed:
apply graphics padding to the msak bounds

NuclearCookie committed:
put back maskMatrix where it belongs

NuclearCookie committed:
Fixed offset in scaled mask

Commit:
Fixed filter offset when container is scaled

NuclearCookie committed:
restored function signature for compatibility

Commit:
Fixed 9 slice bitmap being rendered with invalid scaling

NuclearCookie commented:
Review status: 0 of 24 files reviewed at latest revision, 5 unresolved discussions.

---

*[format/swf/lite/MovieClip.hx, line 139 at r11](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-K_j0EnXsTSxRH646Amy:-K_j0EnXsTSxRH646Amz:bmetkql) ([raw file](https://github.com/fishingcactus/openfl/blob/155f8c632069518db6cfee8dfcc22985e132ea7c/format/swf/lite/MovieClip.hx#L139)):*
> if (!__useSeparateRenderScaleTransform) {
> 			
> 			for( i in 0 ... numChildren ){
> 				var child = getChildAt(i);
> 				child.__useSeparateRenderScaleTransform = false;
> 			}
> 			
> 		}
>  

What about adding a child using AddChild ? Don't we have to do the same check at that moment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/118)*
<!-- Sent from Reviewable.io -->

Commit:
Bounds getter functions are not expanding the provided Rectangle anymore (fixes issues with rectangles retrieved from the pool)

Commit:
Reset blend mode when rendering to bitmap

NuclearCookie committed:
cached renderscale variables

NuclearCookie committed:
Updated mask update condition. Use existing update flags

NuclearCookie committed:
If a color transform is passed to bitmapdata, make sure it is a default one

Commit:
Factorized bitmap caching code + fixed mask matrix

Commit:
Merge pull request #130 from FishingCactus/fix/mask_bitmap_transform

Fix/mask bitmap transform

Commit:
Fixed BitmapData width and height getters

Commit:
Merge remote-tracking branch 'origin/develop' into feature-reduced_hierarchy_transform_update

Commit:
Disable separate render scale transform on child when parent doesn't use it

rams3s commented:
Review status: 0 of 27 files reviewed at latest revision, 5 unresolved discussions.

---

*[format/swf/lite/MovieClip.hx, line 716 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-KZ7YFfRCclnsHq5FDmu-r1-716:-KaC6y2w-9YV91rkWAi5:b-896fix) ([raw file](https://github.com/fishingcactus/openfl/blob/ee1fed5116e51c3d60a686020003cbc516600955/format/swf/lite/MovieClip.hx#L716)):*
<details><summary><i>Previously, NuclearCookie (Pieter Vantorre) wrote…</i></summary><blockquote>

you adressed these todos
</blockquote></details>

Done.

---

*[format/swf/lite/MovieClip.hx, line 139 at r11](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-K_j0EnXsTSxRH646Amy:-KaBqTf6KRWyC3iraVqu:b-hm308m) ([raw file](https://github.com/fishingcactus/openfl/blob/155f8c632069518db6cfee8dfcc22985e132ea7c/format/swf/lite/MovieClip.hx#L139)):*
<details><summary><i>Previously, NuclearCookie (Pieter Vantorre) wrote…</i></summary><blockquote>

> if (!__useSeparateRenderScaleTransform) {
> 			
> 			for( i in 0 ... numChildren ){
> 				var child = getChildAt(i);
> 				child.__useSeparateRenderScaleTransform = false;
> 			}
> 			
> 		}
>  

What about adding a child using AddChild ? Don't we have to do the same check at that moment?

</blockquote></details>

Shouldn't happen (__useSeparateRenderScaleTransform is false for Bitmaps and 9 slice shapes only) but moved this code to addChild for safety. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/118)*
<!-- Sent from Reviewable.io -->
